### PR TITLE
Remove make strings children with null mask

### LINKED ISSUE #8580: [FEA] Remove make_strings_children_with_null_mask utility
The `cudf::strings::detail::make_strings_children_with_null_mask` requires an extra byte array the size of the output column which it uses to convert to a bitmask for the output column. It is used in only `join_list_elements` and `interleave_list_entries` right now. 

The extra byte array is not required since the bitmask can be created directly using the `cudf::detail::valid_if` utility instead. Or a custom kernel could be used to build the offsets and the bitmask to avoid the extra kernel call.

### PR DESCRIPTION
**Pull request by davidwendt**
Closes #8580 

The `cudf::strings::detail::make_strings_children_with_null_mask` utility was created temporarily to help build the output column validities bitmask for `join_lists_elements` (for strings) and `lists::interleave_columns` (for strings). But it used a temporary `int8_t` device vector to hold single-bit values. It would then convert the `int8` column into a bitmask with a kernel call. This PR removes the utility in favor of executing a kernel using the `cudf::detail::valid_if` utility to build the bitmask directly without requiring a temporary buffer. Removing the temporary buffer from the `join_list_elements` strings API was not difficult. The temporary buffer is still used in the `lists::interleave_columns` for now.

 A follow on PR should change this to utilize the output bitmask and directly set the bits rather than using a temporary `int8` buffer that gets converted to a bitmask. This approach could also be used in the `join_lists_element` to ultimately avoid the `valid_if` call.

Removing this utility simplifies the code a bit and should speed up compiling any source file that includes `cudf/strings/detail/utilities.cuh` (~160 files right now).

### CODE REVIEW

File: cpp/src/lists/interleave_columns.cu [hunk 1]
```diff
@@ -172,29 +172,24 @@ struct interleave_list_entries_fn {
...
-                                 std::move(offsets_column),
-                                 std::move(chars_column),
-                                 null_count,
-                                 std::move(null_mask),
-                                 stream,
-                                 mr);
-    }
+    auto validities =
+      rmm::device_uvector<int8_t>(data_has_null_mask ? num_output_entries : 0, stream);
+    comp_fn.d_validities = validities.data();
```
Comment (robertmaynard) [inline cpp/src/lists/interleave_columns.cu:180]: I have concerns about what happens when `data_has_null_mask` is false. The pointer returned by `validities.data()` can't be used or compared against `nullptr`. 

Are we sure in that case that `comp_fn.d_validities` isn't used at all?

Comment (davidwendt) [reply cpp/src/lists/interleave_columns.cu:180]: Yes. The previous implementation already handled that case
This is just moving `validities` allocate from the now-removed utility to here.
https://github.com/rapidsai/cudf/pull/8830/files#diff-8d4b25b181461a330b69fb2f2dc66313e81b08c3dc7a45b2aa2560f50629c9c1L243

Comment (robertmaynard) [reply cpp/src/lists/interleave_columns.cu:180]: Didn't the previous version only call the function when `data_has_null_mask` resolved to true, and now we call the new function even when that is false?

Comment (davidwendt) [reply cpp/src/lists/interleave_columns.cu:180]: Yes, but both paths used the same functor `comp_fn`.

**Review by robertmaynard — APPROVED** [review]

**Review by codereport — APPROVED** [review]
lgtm :+1:

**Comment by davidwendt** [comment]
@gpucibot merge

**Pull request merged**